src/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -97,33 +65,26 @@ public void OnOperationStartHandler(ICallable operation, IApplyData arguments)
         /// </summary>
         public void OnOperationEndHandler(ICallable operation, IApplyData result)
         {
-            this.currentDepth--;
+            if (this.operations.Count <= 1) return;
 
-            // If `compositeTracer` is not null, handle the incoming operation recursively.
-            if (this.compositeTracer != null)
-            {
-                // If the current operation is the composite operation we started with, append
-                // the operations traced out by the `compositeTracer` to the current list of
-                // operations and reset.
-                if (operation == this.currCompositeOp)
-                {
-                    this.AddCompositeOperations();
-                    this.currCompositeOp = null;
-                    this.compositeTracer = null;
-                    return;
-                }
-
-                // Pass operations down to it for handling
-                this.compositeTracer.OnOperationEndHandler(operation, result);
-                return;
-            }
+            var currentOperation = this.operations.Pop();
+            if (currentOperation == null) return;
 
-            // Add latest parsed operation to list of operations, if not null
-            if (this.currentOperation != null)
-            {
-                this.operations.Add(this.currentOperation);
-                this.currentOperation = null;
-            }
+            Operation? parentOp = this.operations.Peek();
+            if (parentOp == null) return;
+
+            // CNOTs are Controlled X under the hood, so we don't need to render the nested CNOT
+            if ((currentOperation.Gate == "X" && currentOperation.IsControlled) &&
+                (parentOp.Gate == "X" && parentOp.IsControlled)) return;
+
+            // Add operation to parent operation's children
+            if (parentOp.Children == null) parentOp.Children = new List<Operation>() { currentOperation };
+            else parentOp.Children = parentOp.Children.Append(currentOperation);
```

Comment:
[nit] May be slightly more readable this way:
```suggestion
            parentOp.Children = parentOp.Children?.Append(currentOperation) ?? new List<Operation>() { currentOperation };
```

Reply:
or, another option:
```suggestion
            parentOp.Children = (parentOp.Children ?? new List<Operation>()).Append(currentOperation);
```

Reply:
@rmshaffer thanks for these suggestions! I love these tips on how to make C# code more compact and neater, so keep em coming 😉

Reply:
There's a subtle issue that may come up here in that `Enumerable.Append(this IEnumerable<TSource>, TSource)` returns an `IEnumerable<TSource>` rather than a collection; I'd suggest using `.ToList` to explicitly save that new enumerable as a list. Even better, you've effectively treated `parentOp.Children` as an _immutable_ list here, since you make a new list from it and save that back to the mutable variable without modifying the list itself. That speaks to using `System.Collections.Immutable.ImmutableList` and its `.Add` method instead.